modules/host/upload.go

```diff
@@ -13,53 +16,447 @@ import (
-	if !exists {
-		h.mu.RUnlock()
-		return errors.New("no record of that file")
+	err = h.considerContract(contractTxn, renterKey, filesize, merkleRoot)
+	h.mu.RUnlock()
+	if err != nil {
+		_ = encoding.WriteObject(conn, err.Error())
+		return errors.New("rejected file contract: " + err.Error())
+	}
+	if err := encoding.WriteObject(conn, modules.AcceptResponse); err != nil {
+		return errors.New("couldn't write acceptance: " + err.Error())
+	}
+
+	// Add collateral to the transaction and send the new transaction.
+	// Currently, collateral is not supported, so the unchanged transaction set
+	// is returned.
+	if err := encoding.WriteObject(conn, unsignedTxnSet); err != nil {
+		return errors.New("couldn't write collateral transaction set: " + err.Error())
 	}
+
+	// The renter will sign the transaction set, agreeing to pay the host.
+	var signedTxnSet []types.Transaction
+	if err := encoding.ReadObject(conn, &signedTxnSet, maxContractLen); err != nil {
+		return errors.New("couldn't read signed transaction set:" + err.Error())
+	}
+
+	// The host will verify that the signed transaction set provided by the
+	// renter is the same transaction set that the host considered previously.
+	if len(signedTxnSet) != len(unsignedTxnSet) {
+		return errors.New("renter sent bad signed transaction set")
+	}
+	for i := range signedTxnSet {
+		if signedTxnSet[i].ID() != unsignedTxnSet[i].ID() {
+			return errors.New("renter sent bad signed transaction set")
+		}
+	}
+
+	// The host now signs the transaction set, confirming the collateral, and
+	// then submits the transaction set to the blockchain.
+	signedTxn, parents := signedTxnSet[len(signedTxnSet)-1], signedTxnSet[:len(signedTxnSet)-1]
+	txnBuilder := h.wallet.RegisterTransaction(signedTxn, parents)
+	signedTxnSet, err = txnBuilder.Sign(true)
+	if err != nil {
+		return err
+	}
+	err = h.tpool.AcceptTransactionSet(signedTxnSet)
+	if err == modules.ErrDuplicateTransactionSet {
+		// This can happen if the host is uploading to itself.
+		err = nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Add this contract to the host's list of obligations.
+	co := &contractObligation{
+		ID:        contractTxn.FileContractID(0),
+		OriginTxn: contractTxn,
+		Path:      filename,
+	}
+	h.mu.Lock()
+	h.addObligation(co)
+	h.mu.Unlock()
+	if err != nil {
+		return err
+	}
+
+	// Send the fully signed and valid transaction set back to the renter.
+	if err := encoding.WriteObject(conn, signedTxnSet); err != nil {
+		return errors.New("couldn't write signed transaction set: " + err.Error())
+	}
+
+	return nil
+}
+
+// managedRPCUpload is an RPC that negotiates a file contract. Under the new
+// scheme, file contracts should not initially hold any data.
+func (h *Host) managedRPCUpload(conn net.Conn) error {
+	// Check that the host has grabbed an address from the wallet.
+	h.mu.RLock()
+	uh := h.settings.UnlockHash
+	h.fileCounter++ // Harmless to increment the file counter in the event of an error.
```

Comment:
does this number get reported to the user anywhere? Or is it just used to name files?